compute/resource_pipeline.go

```diff
@@ -0,0 +1,231 @@
+	Continuous          bool              `json:"continuous,omitempty"`
+	AllowDuplicateNames bool              `json:"allow_duplicate_names,omitempty"`
+}
+
+type createPipelineResponse struct {
+	PipelineID string `json:"pipeline_id"`
+}
+
+// PipelineState ...
+type PipelineState string
+
+// Constants for PipelineStates
+const (
+	StateDeploying  PipelineState = "DEPLOYING"
+	StateStarting   PipelineState = "STARTING"
+	StateRunning    PipelineState = "RUNNING"
+	StateStopping   PipelineState = "STOPPPING"
+	StateDeleted    PipelineState = "DELETED"
+	StateRecovering PipelineState = "RECOVERING"
+	StateFailed     PipelineState = "FAILED"
+	StateResetting  PipelineState = "RESETTING"
+	StateIdle       PipelineState = "IDLE"
+)
+
+// PipelineHealthStatus ...
+type PipelineHealthStatus string
+
+// Constants for PipelineHealthStatus
+const (
+	HealthStatusHealthy   PipelineHealthStatus = "HEALTHY"
+	HealthStatusUnhealthy PipelineHealthStatus = "UNHEALTHY"
+)
+
+type pipelineInfo struct {
+	PipelineID string                `json:"pipeline_id"`
+	Spec       *pipelineSpec         `json:"spec"`
+	State      *PipelineState        `json:"state"`
+	Cause      string                `json:"cause"`
+	ClusterID  string                `json:"cluster_id"`
+	Name       string                `json:"name"`
+	Health     *PipelineHealthStatus `json:"health"`
+}
+
+type pipelinesAPI struct {
+	client *common.DatabricksClient
+	ctx    context.Context
+}
+
+func newPipelinesAPI(ctx context.Context, m interface{}) pipelinesAPI {
+	return pipelinesAPI{m.(*common.DatabricksClient), ctx}
+}
+
+func (a pipelinesAPI) create(s pipelineSpec, allowDuplicateNames bool) (string, error) {
+	var id createPipelineResponse
+	err := a.client.Post(a.ctx, "/pipelines", s, &id)
+	return id.PipelineID, err
+}
+
+func (a pipelinesAPI) read(id string) (p pipelineInfo, err error) {
+	err = a.client.Get(a.ctx, "/pipelines/"+id, nil, &p)
+	return
+}
+
+func (a pipelinesAPI) update(id string, s pipelineSpec, allowDuplicateNames bool) error {
+	return a.client.Put(a.ctx, "/pipelines/"+id, s)
+}
+
+func (a pipelinesAPI) delete(id string) error {
+	return a.client.Delete(a.ctx, "/pipelines/"+id, map[string]string{})
+}
+
+func (a pipelinesAPI) waitForState(id string, timeout time.Duration, desiredState PipelineState) error {
+	return resource.RetryContext(a.ctx, timeout,
+		func() *resource.RetryError {
+			i, err := a.read(id)
+			if err != nil {
+				return resource.NonRetryableError(err)
+			}
+			state := *i.State
+			if state == desiredState {
+				return nil
+			}
+			return resource.RetryableError(fmt.Errorf("Pipeline %s is in state %s, not yet in state %s", id, state, desiredState))
+		})
+}
+
+func (a pipelinesAPI) waitForDeleted(id string, timeout time.Duration) error {
+	return resource.RetryContext(a.ctx, timeout,
+		func() *resource.RetryError {
+			i, err := a.read(id)
+			if err != nil {
+				if e, ok := err.(common.APIError); ok && e.IsMissing() {
```

Comment:
technically, we can simplify this the following way:

* move `IsMissing` check to `delete` method
* just return error as non-retryable.

when you return a retryable error, please also log something, as this will tell the one debugging it with TF_LOG=DEBUG that something is still working.

Reply:
Oh interesting, that message is not logged?

Reply:
so that message is logged as http request on DEBUG level, but i think every state await should be logged as INFO as well.